core/src/main/java/io/grpc/internal/InternalHandlerRegistry.java

```diff
@@ -31,31 +31,49 @@
 
 package io.grpc.internal;
 
+import io.grpc.ExperimentalApi;
 import io.grpc.ServerMethodDefinition;
 import io.grpc.ServerServiceDefinition;
 
+import java.util.ArrayList;
 import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
 
 import javax.annotation.Nullable;
 
 final class InternalHandlerRegistry {
+
+  private final List<ServerServiceDefinition> services;
   private final Map<String, ServerMethodDefinition<?, ?>> methods;
 
-  private InternalHandlerRegistry(Map<String, ServerMethodDefinition<?, ?>> methods) {
+  private InternalHandlerRegistry(
+      List<ServerServiceDefinition> services, Map<String, ServerMethodDefinition<?, ?>> methods) {
+    this.services = services;
     this.methods = methods;
   }
 
+  /**
+   * Returns the service definitions in this registry.
+   */
+  @ExperimentalApi("https://github.com/grpc/grpc-java/issues/2222")
+  public List<ServerServiceDefinition> getServices() {
+    return services;
+  }
+
   @Nullable
   ServerMethodDefinition<?, ?> lookupMethod(String methodName) {
     return methods.get(methodName);
   }
 
   static class Builder {
+
     // Store per-service first, to make sure services are added/replaced atomically.
-    private final HashMap<String, ServerServiceDefinition> services =
-        new HashMap<String, ServerServiceDefinition>();
+    private final SortedMap<String, ServerServiceDefinition> services =
```

Comment:
Optional, but I would have picked LinkedHashMap.  Requiring that they be sorted is asking a lot.  Just being consistent is good enough.